QATCH/QModel/models/qmodel_v7/dp_decode.py

```
@@ -0,0 +1,378 @@
+) -> Optional[Dict[str, Candidate]]:
+    """Exact decode of the blended (seconds + span-fraction) objective.
+
+    The span-fraction denominators depend only on the first and last placed
+    candidates, so we enumerate those pairs, fix the span, and run an exact
+    DP for each. Returns the best configuration under consistent re-scoring,
+    or None if no admissible path exists for any pair."""
+    first, last = placeable[0], placeable[-1]
+    best_chosen: Optional[Dict[str, Candidate]] = None
+    best_score = -1e18
+    for f in cand[first]:
+        for l in cand[last]:
```

Ruff (0.15.20)

[error] Ambiguous variable name: `l`

(E741)